docs/stability.md

```diff
@@ -0,0 +1,39 @@
+---
+title: API Stability
+sort_rank: 8
+---
+
+# API Stability Guarantees
+
+Prometheus promises API stability within a major version, and strives to avoid
+breaking changes for key features. Some features which are cosmetic, still
+under development, or depend on 3rd party services are not covered by this.
+
+Things considered stable for 2.x:
+
+* The query language and data model
+* Alerting and recording rules
+* The ingestion exposition format
+* JSON HTTP API (used by dashboards and UIs)
+* Configuration file format (minus the service discovery remote read/write, see below)
+* Rule/alert file format
+* Console template syntax and semantics
+
+Things considered unstable for 2.x:
+
+* The [`holt_winters` PromQL function](https://github.com/prometheus/prometheus/issues/2458)
+* Remote read, remote write and the remote read endpoint
+* GRPC APIs
+* Flag meaning and names
```

Comment:
This was previously declared stable in 1.x, no? What is our reasoning behind this?

Reply:
Only the names were stable in 1.x, and it caused oddities. This wording enables us to hard break a user if there's a flag change they really need to think about.

Reply:
I'm happy with this decision. Just wanted to clarify.

Reply:
If there is a flag change they really have to think about, we have to release the major version. I'm against doing our own versioning but following semver, so that version numbers have some meaning.

Reply:
I'm specifically thinking about `--external-url` and `--route-prefix` as they lead to frequent confusion. If we want to be more detailed, I would like to exclude those from the stability guarantees. I can see how flags, is something that people would likely complain about if they break and not allow the prometheus server to start in a consecutive release.

Reply:
What if it's a minor flag that most users don't touch, but matters in some obscure use case? For example if anyone was using the flags for talking out to influx we silently started dropping their data on the floor at some point during 1.x. That is not good behaviour, breaking them explicitly would have been better than breaking them silently.

Reply:
> I can see how flags, is something that people would likely complain about if they break and not allow the prometheus server to start in a consecutive release.

That is something some users always complain about, as they don't understand why a hard break is a necessary sometimes. They would prefer the binary to start but now do undesirable things - which they'll likely also blame us for once they discover that. There's no way to win for us, and I'd prefer to be explicit about breakage.

Reply:
So what flags are you worried about. The influx flags were related to a feature which was experimental to begin with. The only change which came up was related to proxy behavior. I believe if we come up with a better implementation, it should be done with a new flag name so existing documentation and blog posts can't contradict the new behavior. That can be introduced without changing existing flag behavior.

Reply:
There's no particular ones concerning me currently, but I'd like to keep our options open. For example I can imagine issues arising around the block durations, which are only theoretically there for testing.

Reply:
So there doesn't exist a reason to exclude all flags from the stability guarantees. The only time it caused an issue was solely for an experimental feature which was already excluded from the stability guarantees any way. Given the tsdb flags are not meant for production usage anyway, I'd be fine explicitely excluding these as you seem to prefer. For anything else, we'd need to bump the major version.

You might have a different reading, but I find the linked document we agreed upon quite clear. The goal is not to follow the semver spec less with more releases, but the opposite. I consider it vastly more important to have a semantically meaningful major version, than to keep options open for hypothetically cases and being afraid of major release bumps. To quote the "actual versioning" section of the document a last time.

> To stay relevant we should not be afraid of breaking. Releasing major versions frequently will help with that.

Either reduce flag changes to the dev related tsdb options, or start a thread on prometheus-dev about your opinion on versioning. I don't have anything else to add here anymore.

Reply:
> So there doesn't exist a reason to exclude all flags from the stability guarantees. 

There is no glaring reason right now today, this is about giving us flexibility for the future which we cannot predict. 2.x will be around for a while, and more flags will be added.

> The only time it caused an issue was solely for an experimental feature which was already excluded from the stability guarantees any way.

Given our wording, it was not permitted to hard break those users - even for an experimental feature - as we guaranteed a Prometheus would always start with old flags.
Just because a feature is marked experimental doesn't mean we should let users silently lose data if we can avoid it.

Permitting silent breakage but not explicit breaking is bizarre to me. @juliusv also previously expressed concern when the implications of the 1.x wording became clear.

> I find the linked document we agreed upon quite clear. 

We never agreed the whole doc to my recollection. It was a focus for discussion that petered out, as typically happens with these sorts of doc.

> I consider it vastly more important to have a semantically meaningful major version, than to keep options open for hypothetically cases and being afraid of major release bumps.

That's your opinion. On the basis that the 6 minor releases of 1.x with breaking changes were minor rather than major, that is not the standing policy of the project. (And all 8 minor releases were breaking by a strict definition).

So while the discussion doc may say "Releasing major versions frequently will help with that.", that's not at all what we did in practice.

> or start a thread on prometheus-dev about your opinion on versioning.

If you believe that we should go full semver then the onus is on you to gain consensus on that, as that is not the effective standing policy.